libs/payments/customer/src/lib/invoice.manager.ts

```diff
@@ -104,9 +123,76 @@ export class InvoiceManager {
     invoice: StripeInvoice,
     ipaddress?: string
   ) {
-    // TODO in M3b: Implement legacy processInvoice as processNonZeroInvoice here
-    // TODO: Add spec
-    console.log(customer, invoice, ipaddress);
+    if (!customer.metadata[STRIPE_CUSTOMER_METADATA.PaypalAgreement]) {
+      throw new StripePayPalAgreementNotFoundError(customer.id);
+    }
+    if (!['draft', 'open'].includes(invoice.status ?? '')) {
+      throw new InvalidInvoiceError();
+    }
+
+    // PayPal allows for idempotent retries on payment attempts to prevent double charging.
+    const paymentAttemptCount = parseInt(
+      invoice?.metadata?.[STRIPE_INVOICE_METADATA.RetryAttempts] ?? '0'
+    );
+    const idempotencyKey = `${invoice.id}-${paymentAttemptCount}`;
+
+    // Charge the customer on PayPal
+    const chargeOptions = {
+      amountInCents: invoice.amount_due,
+      billingAgreementId:
+        customer.metadata[STRIPE_CUSTOMER_METADATA.PaypalAgreement],
+      invoiceNumber: invoice.id,
+      currencyCode: invoice.currency,
+      idempotencyKey,
+      ...(ipaddress && { ipaddress }),
+      ...(invoice.tax && { taxAmountInCents: invoice.tax }),
+    } satisfies ChargeOptions;
+    let paypalCharge: ChargeResponse;
+    try {
+      [paypalCharge] = await Promise.all([
+        this.paypalClient.chargeCustomer(chargeOptions),
+        this.stripeClient.invoicesFinalizeInvoice(invoice.id),
+      ]);
+    } catch (error) {
+      if (PayPalClientError.hasPayPalNVPError(error)) {
+        PayPalClientError.throwPaypalCodeError(error);
```

Comment:
It looks like if we fail here, we never increment the attempt count. I believe that may cause all future retries to fail, since we'd always be attempting the same attempt count consumed by the first try's idempotency key.

Reply:
Honestly, this will all likely just go away with some of the changes I'm introducing during the refactor. I'm changing the current front-end polling logic to allow for failures by having failed calls increment the retry count by one.